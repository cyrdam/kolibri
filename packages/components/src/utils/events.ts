// TODO: Should be synchronized with enums/events.ts
export type KoliBriEventType =
	| 'blur'
	| 'change'
	| 'changePage'
	| 'changePageSize'
	| 'click'
	| 'close'
	| 'focus'
	| 'input'
	| 'mousedown'
	| 'reset'
	| 'selection-change' // todo remove
	| 'selectionChange'
	| 'sort'
	| 'submit'
	| 'toggle';

export function stopPropagation(event: Event): void {
	event.stopImmediatePropagation();
	event.stopPropagation();
}

function createKoliBriEvent<T>(type: KoliBriEventType, detail: T): CustomEvent {
	const event = new CustomEvent(`kol-${type}`, {
		bubbles: true,
		cancelable: true,
		composed: true,
		detail: detail,
	});
	return event;
}

function dispatchKoliBriEvent<T>(target: EventTarget, type: KoliBriEventType, detail?: T): boolean {
	const dispatch = target.dispatchEvent(createKoliBriEvent(type, detail));
	return dispatch;
}

/**
 * @deprecated
 */
export function tryToDispatchKoliBriEvent<T>(type: KoliBriEventType, target?: EventTarget, detail?: T): void {
	target && dispatchKoliBriEvent(target, type, detail);
}

export function dispatchDomEvent(target: HTMLElement, type: KoliBriEventType) {
	target.dispatchEvent(new Event(type, { bubbles: true, composed: true }));
}

export function preventDefaultAndStopPropagation(event: Event) {
	event.preventDefault();
	stopPropagation(event);
}
