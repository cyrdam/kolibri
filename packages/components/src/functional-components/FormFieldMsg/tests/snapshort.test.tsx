import { h } from '@stencil/core';
import FormFieldMsgFc from '../FormFieldMsg';
import { renderFunctionalComponentToSpecPage } from '../../../utils/testing';
import type { InternMsgPropType } from '../../../schema';

describe('FormFieldMsgFc', () => {
	it('should render with all props', async () => {
		const alert = true;
		const msg: InternMsgPropType = { description: 'This is an error message' };
		const id = 'test-id';
		const classNames = 'custom-class';
		const page = await renderFunctionalComponentToSpecPage(() => <FormFieldMsgFc alert={alert} msg={msg} id={id} class={classNames} />);

		expect(page.root).toMatchSnapshot();
	});

	it('should set the correct id', async () => {
		const id = 'test-id';
		const page = await renderFunctionalComponentToSpecPage(() => <FormFieldMsgFc id={id} />);

		expect(page.root?.id).toBe(`${id}-msg`);
	});

	it('should set the correct class names', async () => {
		const classNames = 'custom-class';
		const page = await renderFunctionalComponentToSpecPage(() => <FormFieldMsgFc id="test-id" class={classNames} />);

		expect(page.root?.className).toContain('error');
		expect(page.root?.className).toContain(classNames);
	});

	it('should handle hideMsg prop correctly', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <FormFieldMsgFc id="test-id" hideMsg={true} />);

		expect(page.root?.className).toContain('visually-hidden');
	});

	it('should render the message correctly', async () => {
		const msg: InternMsgPropType = { description: 'This is an error message' };
		const page = await renderFunctionalComponentToSpecPage(() => <FormFieldMsgFc id="test-id" msg={msg} />);

		expect(page.root?.textContent).toBe(msg.description);
	});
});
