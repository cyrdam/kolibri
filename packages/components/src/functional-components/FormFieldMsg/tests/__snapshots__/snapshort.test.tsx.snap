// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldMsgFc should render with all props 1`] = `
<div aria-hidden="true" class="custom-class kol-alert kol-alert--error kol-alert--msg kol-form-field__msg" description="This is an error message" id="test-id-msg" role="alert">
  <div class="kol-alert__container">
    <kol-icon _icons="codicon codicon-error" _label="kol-error" class="kol-alert__heading-icon"></kol-icon>
    <div class="kol-alert__container-content">
      <div class="kol-alert__content">
        This is an error message
      </div>
    </div>
  </div>
</div>
`;
